tests/ctests/test_netplan_misc.c

```diff
@@ -177,6 +189,107 @@ test_netplan_netdef_get_output_filename_invalid_backend(__unused void** state)
     assert_int_equal(ret, 0);
 }
 
+void
+test_netplan_netdef_write_yaml(__unused void** state)
+{
+    const char* yaml =
+        "network:\n"
+        "  version: 2\n"
+        "  ethernets:\n"
+        "    eth0:\n"
+        "      dhcp4: true";
+
+    NetplanState* np_state = load_string_to_netplan_state(yaml);
+    NetplanNetDefinition* interface = netplan_state_get_netdef(np_state, "eth0");
+
+    char template[] = "/tmp/netplan.XXXXXX";
+    // no need to free() rootdir, as it will modify the template[] buffer
+    char *rootdir = mkdtemp(template);
+    char etc[24] = {};
+    char etc_netplan[32] = {};
+    snprintf(etc, 24, "%s/etc", rootdir);
+    snprintf(etc_netplan, 32, "%s/netplan", etc);
+    mkdir(etc, 0770);
+    mkdir(etc_netplan, 0770);
+
+    /* Check API call */
+    NetplanError* err = NULL;
+    assert_true(netplan_netdef_write_yaml(np_state, interface, rootdir, &err));
+    assert_true(err == NULL);
+
+    /* Check file exists */
+    struct stat st = {0};
+    char output_yaml[53] = {};
+    snprintf(output_yaml, 53, "%s/10-netplan-eth0.yaml", etc_netplan);
+    assert_true(stat(output_yaml, &st) == 0);
+
+    /* Check file contents */
+    FILE *fd = fopen(output_yaml, "r");
+    char file_buffer[600];
+    assert_true(fread(file_buffer, 1, strlen(yaml), fd) > 0);
+    assert_string_equal(yaml, file_buffer);
+
+    /* Cleanup */
+    netplan_state_clear(&np_state);
+    fclose(fd);
+    remove(output_yaml);
+    rmdir(etc_netplan);
+    rmdir(etc);
+    rmdir(rootdir);
+}
+
+void
+test_netplan_netdef_write_yaml_90NM(__unused void** state)
+{
+    const char* yaml =
+        "network:\n"
+        "  version: 2\n"
+        //"  renderer: NetworkManager\n" //FIXME: renderer get's eaten by the API call...
+        "  ethernets:\n"
+        "    eth0:\n"
+        "      dhcp4: true\n"
+        "      networkmanager:\n"
+        "        uuid: \"990548be-01ed-42d7-9f9f-cd4966b25c08\"";
+
+    NetplanState* np_state = load_string_to_netplan_state(yaml);
+    NetplanNetDefinition* interface = netplan_state_get_netdef(np_state, "eth0");
+
+    char template[] = "/tmp/netplan.XXXXXX";
+    // no need to free() rootdir, as it will modify the template[] buffer
+    char *rootdir = mkdtemp(template);
+    char etc[24] = {};
+    char etc_netplan[32] = {};
+    snprintf(etc, 24, "%s/etc", rootdir);
+    snprintf(etc_netplan, 32, "%s/netplan", etc);
+    mkdir(etc, 0770);
+    mkdir(etc_netplan, 0770);
+
+    /* Check API call */
+    NetplanError* err = NULL;
+    assert_true(netplan_netdef_write_yaml(np_state, interface, rootdir, &err));
+    assert_true(err == NULL);
+
+    /* Check file exists */
+    struct stat st = {0};
+    char output_yaml[80] = {};
+    snprintf(output_yaml, 80, "%s/90-NM-990548be-01ed-42d7-9f9f-cd4966b25c08.yaml", etc_netplan);
+    assert_true(stat(output_yaml, &st) == 0);
+
+    /* Check file contents */
+    FILE *fd = fopen(output_yaml, "r");
+    char file_buffer[600];
```

Comment:
The unittest is failing with ASAN because this buffer is not being zeroed (on the previous test as well but it's not failing...). There is garbage in the end of string (see below). Initializing it with zeros (` = {0}`) fixes the test. The memory leak happens because the `assert_string_equal` assertion prevents `np_state` to be cleared.

```
[  ERROR   ] --- "network:
  version: 2
  ethernets:
    eth0:
      dhcp4: true
      networkmanager:
        uuid: "990548be-01ed-42d7-9f9f-cd4966b25c08"" != "network:
  version: 2
  ethernets:
    eth0:
      dhcp4: true
      networkmanager:
        uuid: "990548be-01ed-42d7-9f9f-cd4966b25c08"IH"
[   LINE   ] --- ../tests/ctests/test_netplan_misc.c:282: error: Failure!
```

Reply:
The RPM build is failing for the same reason (see the end of the buffer below). But there both tests are failing without ASAN, which is interesting.

```
[  ERROR   ] --- "network:
  version: 2
  ethernets:
    eth0:
      dhcp4: true
      networkmanager:
        uuid: "990548be-01ed-42d7-9f9f-cd4966b25c08"" != "network:
  version: 2
  ethernets:
    eth0:
      dhcp4: true
      networkmanager:
        uuid: "990548be-01ed-42d7-9f9f-cd4966b25c08"oÏwBV"
[   LINE   ] --- ../tests/ctests/test_netplan_misc.c:282: error: Failure!
```